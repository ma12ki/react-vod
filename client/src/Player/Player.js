import React, { PropTypes } from 'react';

export const Player = (props) => {
    const { id, ext } = props;
    const type = getType(ext);
    const url = `http://localhost:3001/play/${id}`;

    return (
        <video controls>
            <source src={url} type={type} />
            Your browser does not support the <code>video</code> element.
        </video>
    );
};

Player.propTypes = {
    id: PropTypes.string.isRequired,
    ext: PropTypes.string.isRequired,
};

const getType = (ext) => {
    switch (ext) {
        case '.mp4':
            return 'video/mp4';
        case '.webm':
            return 'video/webm';
        default:
            throw new Error(`Unknown format: ${ext}`);
    }
};

export default Player;
