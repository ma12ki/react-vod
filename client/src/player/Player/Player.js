import React from 'react';
import PropTypes from 'prop-types';

import { apiUrl } from '../../config';

export const Player = (props) => {
    const { id, ext } = props;
    const type = getType(ext);
    const url = `${apiUrl}play/${id}`;

    return (
        <video controls preload='metadata'>
            <source src={url} type={type} />
            Your browser does not support the <code>video</code> element.
        </video>
    );
};

Player.propTypes = {
    id: PropTypes.string.isRequired,
    ext: PropTypes.string.isRequired,
};

const getType = (ext) => {
    switch (ext) {
        case '.mp4':
            return 'video/mp4';
        case '.webm':
            return 'video/webm';
        default:
            throw new Error(`Unknown format: ${ext}`);
    }
};

export default Player;
