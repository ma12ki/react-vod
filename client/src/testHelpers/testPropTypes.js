import React from 'react';

// this has some limitations:
// https://stackoverflow.com/questions/41916992/why-does-my-jest-test-of-react-proptypes-break-when-using-multiple-unacceptable/41916993#41916993
export const testPropTypes = (component, propName, acceptableValues, unacceptableValues, otherProps) => {
    console.error = jest.fn();
    const _test = (testValues, expectError) => {
        for (let propValue of testValues) {
            console.error.mockClear();
            React.createElement(component, {...otherProps, [propName]: propValue});
            try {
                expect(console.error).toHaveBeenCalledTimes(expectError ? 1 : 0);
            } catch (err) {
                console.log(`${propName} expected to ${expectError ? '' : 'NOT'} fail for value ${propValue}`);
                throw err;
            }
        }
    };
    _test(acceptableValues || [], false);
    _test(unacceptableValues || [], true);
};
