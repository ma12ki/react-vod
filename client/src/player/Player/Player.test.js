import React from 'react';
import { shallow } from 'enzyme';

import { testPropTypes } from '../../testHelpers';
import { Player } from './Player';

jest.mock('../../config', () => {
    return {
        apiUrl: 'BEST_API'
    };
});

describe('Player', () => {

    beforeEach(() => jest.clearAllMocks());

    it('matches the snapshot', () => {
        const id = 'omg';
        const ext = '.mp4';
        const wrapper = shallow(<Player id={id} ext={ext} />);

        expect(wrapper).toMatchSnapshot();
    });

    it('requires "id" string prop', () => {
        testPropTypes(
            Player,
            'id',
            ['rofl', 'xD'],
            [123, null, undefined],
            {
                ext: '.mp4'
            }
        );
    });

    it('requires "ext" string prop', () => {
        testPropTypes(
            Player,
            'ext',
            ['.mp4', '.webm'],
            [123, null, undefined],
            {
                ext: '.mp4'
            }
        );
    });

});