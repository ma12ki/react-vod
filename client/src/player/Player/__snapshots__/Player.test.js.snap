// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player matches the snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="ar16_9"
>
    <video
        controls={true}
        preload="metadata"
    >
        <source
            src="BEST_APIplay/omg"
            type="video/mp4"
        />
        Your browser does not support the 
        <code>
            video
        </code>
         element.
    </video>
</div>,
  "nodes": Array [
    <div
      className="ar16_9"
>
      <video
            controls={true}
            preload="metadata"
      >
            <source
                  src="BEST_APIplay/omg"
                  type="video/mp4"
            />
            Your browser does not support the 
            <code>
                  video
            </code>
             element.
      </video>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Unknown
        ext=".mp4"
        id="omg"
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "ext": ".mp4",
          "id": "omg",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="ar16_9"
>
          <video
                    controls={true}
                    preload="metadata"
          >
                    <source
                              src="BEST_APIplay/omg"
                              type="video/mp4"
                    />
                    Your browser does not support the 
                    <code>
                              video
                    </code>
                     element.
          </video>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="ar16_9"
>
          <video
                    controls={true}
                    preload="metadata"
          >
                    <source
                              src="BEST_APIplay/omg"
                              type="video/mp4"
                    />
                    Your browser does not support the 
                    <code>
                              video
                    </code>
                     element.
          </video>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Unknown
    ext=".mp4"
    id="omg"
/>,
}
`;
